### PR TITLE
General access control

### PR DESCRIPTION
**Pull request by fbielejec**
# Description

 This is a shot at making access control a go-to contract for checking privileges.

First of all the Owner role is removed (in favour of Admin role) and this is a role for all things-destructive. 
Secondly instead of proliferating roles there is now a Custom role that accepts and identifier which can have an application-specifc semantic meaning, e.g. `LiquidityProvider`, `Minter` or  `Burner` are examples of such roles.

### CODE REVIEW

**Comment by krzysztofziobro** [comment]
Question: Not really concerning this PR, but do I understand correctly, that `grant_role` can only be used by admin of the `AccessControl` contract? Wouldn't we want it to also be available for `Admin` of the specific contract? Of course we cannot grant `Instantiator` this way, but for other roles it should be pretty straightforward.

File: contracts/access_control/lib.rs [hunk 1]
```diff
@@ -217,6 +210,27 @@ mod access_control {
...
+                    .is_ok(),
+                "custom grant_role should work"
+            );
+
+            assert!(
+                access_control.has_role(
+                    bob,
+                    Role::Custom(contract_address, [0x43, 0x53, 0x54, 0x4D])
+                ),
+                "charlie should have a custom role"
```
Comment (krzysztofziobro) [inline contracts/access_control/lib.rs:231]:
```suggestion
                "bob should have a custom role"
```

**Comment by fbielejec** [comment]
> Question: Not really concerning this PR, but do I understand correctly, that `grant_role` can only be used by admin of the `AccessControl` contract? Wouldn't we want it to also be available for `Admin(AccountId)` of the contract with address `AccountId`? Of course we cannot grant `Instantiator` this way, but for other roles it should be pretty straightforward.

My concern is this would start proliferating too much - and Admin(AccountId) could give it to another AccountId, at which point it could also give it to another account and everybody is admin all of a sudden - not to mention that now you need to handle who can remove these admin right form whom - all of which seems like an overkill - at least I haven't seen a need for a more complex or fine grained access control

**Comment by krzysztofziobro** [comment]
I guess that as long as one instance of the `AccessContract` is responsible for managing roles in one "project", then that does not really matter, as it's `Admin` is probably the account controlled by the same user(s) as `Admin` of the `AccessControl`.

It's probably the way that we want it to be used anyway, so yeah, probably not worth it.

**Review by krzysztofziobro — APPROVED** [review]
Looks good

File: contracts/button/lib.rs [hunk 2]
```diff
@@ -241,26 +241,42 @@ pub mod button_game {
...
+        /// Can only be called by the contract admin
+        #[ink(message)]
+        pub fn set_button_lifetime(
+            &mut self,
+            new_button_lifetime: BlockNumber,
+        ) -> ButtonResult<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            let required_role = Role::Admin(this);
+            ButtonGame::check_role(&self.access_control, caller, required_role)?;
```
[the remark this answers is not in the document]
Comment (fbielejec) [reply contracts/button/lib.rs:267]: yeah, nice idea - a decorator macro for AC - will put it into the TODO list!

**Pull request merged**